source/agora/consensus/protocol/Nominator.d

```diff
@@ -0,0 +1,373 @@
+
+        log.info("Externalized transaction set at {}: {}", slot_idx, tx_set);
+        if (!this.ledger.txSetExternalized(tx_set))
+            assert(0);
+    }
+
+    ///
+    public override SCPQuorumSetPtr getQSet (ref const(StellarHash) qSetHash)
+    {
+        logCall("{} {}", this.key_pair.address.prettify);
+        if (auto scp_quroum = qSetHash in this.quorum_set)
+            return *scp_quroum;
+
+        return SCPQuorumSetPtr.init;
+    }
+
+    ///
+    public override void emitEnvelope (ref const(SCPEnvelope) envelope)
+    {
+        logCall("{} {}", this.key_pair.address.prettify);
+        scope (failure) assert(0);
+
+        foreach (key, node; this.peers)
+        {
+            // note: cannot deal with const parameter types in the API
+            auto env = cast()envelope;
+
+            // note: several error-cases not considered here yet:
+            // A) request failure after N attepts => we might have to retry,
+            // but exactly how much time do we have before the next round??
+            // B) Node rejects the envelope. Possible in circular scenarios,
+            // e.g. A => B => C => A (A rejects the envelope because it sent it first)
+            node.sendEnvelope(env);
+        }
+    }
+
+    ///
+    public override Value combineCandidates (uint64_t slot_idx,
+        ref const(set!Value) candidates)
+    {
+        logCall("{} {}", this.key_pair.address.prettify);
+        scope (failure) assert(0);
+
+        foreach (ref const(Value) candidate; candidates)
+        {
+            auto tx_set = deserializeFull!(Set!Transaction)(
+                cast(ubyte[])candidate[]);
+
+            if (auto msg = this.ledger.validateTxSet(tx_set))
+            {
+                log.error("combineCandidates(): Invalid tx set: {}", msg);
+                continue;
+            }
+            else
+            {
+                log.info("combineCandidates: {}", slot_idx);
+            }
+
+            // todo: currently we just pick the first of the candidate values,
+            // but we should ideally pick tx's out of the combined set
+            return tx_set.serializeFull().toVec();
+        }
+
+        assert(0);  // should not reach here
+    }
+
+    ///
+    public override void setupTimer (ulong slotIndex, int timerID,
+        milliseconds timeout, CPPDelegate!SCPCallback* cb)
+    {
+        scope (failure) assert(0);
+        logCall("{} {}", this.key_pair.address.prettify);
+
+        if (cb is null || timeout == 0)
+        {
+            this.timers.remove(timerID);
+            return;
+        }
+
+        this.timers.put(timerID);
+
+        this.taskman.runTask(
+        {
+            this.taskman.wait(timeout.msecs);
+            if (timerID !in this.timers)
+                return;
+
+            callCPPDelegate(cb);
+        });
+    }
+}
+
+/// workaround for logger not being nothrow
+private void logCall (T...)(lazy T args, string func = __FUNCTION__) nothrow
```

Comment:
Can we make a `nothrow` wrapper in Agora an call it a day ?

Reply:
Like a wrapper around `Logger` with an `alias logger this` inside?

Reply:
https://github.com/bpfkorea/agora/pull/464